File: springboot/junit5/src/main/java/com/example/demo/model/ArticleDto.java
package com.example.demo.model;

import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import javax.validation.constraints.NotBlank;
import java.util.*;
import java.util.stream.Collectors;

public class ArticleDto {

    @Getter
    @NoArgsConstructor
    public static class CreateReq {

        @NotBlank(message = "제목 필수 입력")
        private String subject;

        @NotBlank(message = "내용 필수 입력")
        private String contents;
        private Set<String> tags = new LinkedHashSet<>();

        @Builder
        public CreateReq(String subject, String contents, Set<String> tags) {
            this.subject = subject;
            this.contents = contents;
            this.tags = tags;
        }

        public Article toEntity(){

            Article article = Article.builder()
                    .subject(this.subject)
                    .contents(this.contents)
                    .build();

            if(tags != null)
                article.initTags(tags);

            return article;
        }
    }

    @Getter
    @NoArgsConstructor
    public static class ModifyReq {

        private String subject;
        private String contents;
        private Set<String> tags = new LinkedHashSet<>();

        @Builder
        public ModifyReq(String subject, String contents, Set<String> tags) {
            this.subject = subject;
            this.contents = contents;
            this.tags = tags;
        }
    }

    @Getter
    @NoArgsConstructor
    public static class Resp {

        private Long idx;
        private String subject;
        private String contents;
        private List<String> tags;

        @Builder
        public Resp(Long idx, String subject, String contents, List<String> tags) {
            this.idx = idx;
            this.subject = subject;
            this.contents = contents;
            this.tags = tags;
        }

        public static Resp of(Article article){

            List<String> tagList = article.getTags().stream()
                    .map(Tag::getTag)
                    .collect(Collectors.toList());

            return Resp.builder()
                    .idx(article.getIdx())
                    .subject(article.getSubject())
                    .contents(article.getContents())
                    .tags(tagList)
                    .build();
        }
    }
}

File: springboot/HibernateLock/src/main/kotlin/com/example/hibernatelock/article/dto/ArticleDto.kt
package com.example.hibernatelock.article.dto

import com.example.hibernatelock.article.entity.Article
import com.example.hibernatelock.article.entity.ArticleDetail
import javax.validation.constraints.NotBlank

class ArticleCreateRequest(

    var subject: String = "",
    var contents: String = "",
) {

    fun toEntity(): Pair<Article, ArticleDetail> {

        val article = Article(
            subject = subject
        )

        val articleDetail = ArticleDetail(
            subject = subject, contents = contents, article = article
        )

        return article to articleDetail
    }
}


class ArticleDto(
    val subject: String,
    val viewCount: Long,
    val version: Int,
) {


    companion object {

        fun of(
            article: Article
        ): ArticleDto = ArticleDto(
            subject = article.subject,
            viewCount = article.viewCount,
            version = article.version,
        )
    }
}

class ArticleDetailDto(
    val subject: String,
    val contents: String,
    val viewCount: Long,
    val articleVersion: Int,
    val detailVersion: Int,
) {

    companion object {

        fun of(
            articleDetail: ArticleDetail
        ): ArticleDetailDto = ArticleDetailDto(
            subject = articleDetail.subject,
            contents = articleDetail.contents,
            viewCount = articleDetail.article.viewCount,
            articleVersion = articleDetail.article.version,
            detailVersion = articleDetail.version,
        )
    }
}

class ArticleUpdateRequest(

    @field:NotBlank
    var subject: String = "",
    @field:NotBlank
    var contents: String = ""
)